### PR TITLE
11.7.1 features gcf ci tests

### PR DESCRIPTION
**Pull request by eranha**
### What does this PR do?
Adds Google function CI tests according to test plan.

### What ticket does this PR close?
Connected to #[relevant GitHub issues, #1811 ]

### CODE REVIEW

File: cucumber/authenticators_gcp/features/authn_gcf.feature [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    And I save my place in the log file
+    When I authenticate with authn-gcp using valid GCF identity token and existing account
+    Then host "test-app" has been authorized by Conjur
+    And I can GET "/secrets/cucumber/variable/test-variable" with authorized user
+    And The following appears in the audit log after my savepoint:
+    """
+    cucumber:host:test-app successfully authenticated with authenticator authn-gcp service cucumber:webservice:conjur/authn-gcp
+    """
+
+  Scenario: Hosts can authenticate with GCP authenticator using ID annotation only and fetch secret
```
Comment (nessiLahav) [inline cucumber/authenticators_gcp/features/authn_gcf.feature:42]: Change ID to the annotation name: `service-account-id`

File: cucumber/authenticators_gcp/features/authn_gcf.feature [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    Given I have a "variable" resource called "test-variable"
+    And I add the secret value "test-secret" to the resource "cucumber:variable:test-variable"
+    And I permit host "test-app" to "execute" it
+    And I set id GCF annotations to host "test-app"
```
Comment (nessiLahav) [inline cucumber/authenticators_gcp/features/authn_gcf.feature:46]: you need to delete all annotations before this step, to make sure you do not have left overs form previous scenarios:
`And I remove all annotations from host "test-app"`

File: cucumber/authenticators_gcp/features/authn_gcf.feature [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    And The following appears in the audit log after my savepoint:
+    """
+    cucumber:host:test-app successfully authenticated with authenticator authn-gcp service cucumber:webservice:conjur/authn-gcp
+    """
+
+  Scenario: Hosts can authenticate with GCP authenticator using ID annotation only and fetch secret
+    Given I have a "variable" resource called "test-variable"
+    And I add the secret value "test-secret" to the resource "cucumber:variable:test-variable"
+    And I permit host "test-app" to "execute" it
+    And I set email GCF annotations to host "test-app"
```
Comment (nessiLahav) [inline cucumber/authenticators_gcp/features/authn_gcf.feature:61]: Why no to use the real annotation name `authn-gcp/servcie-account-email` ? i think it is more clear

File: cucumber/authenticators_gcp/features/authn_gcf.feature [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    And I save my place in the log file
+    When I authenticate with authn-gcp using valid GCF identity token and existing account
+    Then host "test-app" has been authorized by Conjur
+    And I can GET "/secrets/cucumber/variable/test-variable" with authorized user
+    And The following appears in the audit log after my savepoint:
+    """
+    cucumber:host:test-app successfully authenticated with authenticator authn-gcp service cucumber:webservice:conjur/authn-gcp
+    """
+
+  Scenario: Hosts can authenticate with GCP authenticator using ID annotation only and fetch secret
```
Comment (nessiLahav) [inline cucumber/authenticators_gcp/features/authn_gcf.feature:57]: change ID to the annotation name `servcie-account-email`

File: cucumber/authenticators_gcp/features/authn_gcf.feature [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    And I set email GCF annotations to host "test-app"
+    And I obtain a valid GCF identity token
+    And I save my place in the log file
+    When I authenticate with authn-gcp using valid GCF identity token and existing account
+    Then host "test-app" has been authorized by Conjur
+    And I can GET "/secrets/cucumber/variable/test-variable" with authorized user
+    And The following appears in the audit log after my savepoint:
+    """
+    cucumber:host:test-app successfully authenticated with authenticator authn-gcp service cucumber:webservice:conjur/authn-gcp
+    """
```
Comment (nessiLahav) [inline cucumber/authenticators_gcp/features/authn_gcf.feature:70]: missing a new line

File: cucumber/authenticators_gcp/features/authn_gcf_token_errors.feature [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    And I set all valid GCF annotations to host "test-app"
+    And I set "authn-gcp/instance-name" annotation to function host "test-app"
+    And I obtain a valid GCF identity token
+    And I save my place in the log file
+    When I authenticate with authn-gcp using valid GCF identity token and existing account
+    Then it is unauthorized
+    And The following appears in the log after my savepoint:
+    """
+    CONJ00068E Resource restriction 'instance-name' does not exists resource in JWT token
+    """
```
Comment (nessiLahav) [inline cucumber/authenticators_gcp/features/authn_gcf_token_errors.feature:93]: missing a new line

File: cucumber/authenticators_gcp/features/authn_gcf.feature [hunk 7]
```diff
@@ -0,0 +1,70 @@
+Feature: GCP Authenticator - Hosts can authenticate with GCP authenticator
```
Comment (nessiLahav) [inline cucumber/authenticators_gcp/features/authn_gcf.feature:1]: I think all feature names should also contain GCF or GCE, @eladkug do yo u agree?

File: cucumber/authenticators_gcp/features/authn_gcf_token_errors.feature [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      - !permit
+        role: !group apps
+        privilege: [ read, authenticate ]
+        resource: !webservice
+    """
+    And I have host "test-app"
+    And I grant group "conjur/authn-gcp/apps" to host "test-app"
+
+  Scenario: Token with service-account-id claim that does not match annotation is denied
+    And I set "authn-gcp/service-account-id" annotation with value: "unknown-service-account-id" to host "test-app"
```
Comment (nessiLahav) [inline cucumber/authenticators_gcp/features/authn_gcf_token_errors.feature:26]: Relevant to the whole PR:
Remove annoations before setting a new one

**Review by nessiLahav — CHANGES_REQUESTED** [review]

File: cucumber/authenticators_gcp/features/authn_gcf.feature (hunk 7 above)
Comment (eladkug) [reply cucumber/authenticators_gcp/features/authn_gcf.feature:1]: Need to add global label @GCP before the first feature(it will give us the possibility in the future to run all gcp tests by one command)
Each feature should declare what it testing GCF or GCE.

File: cucumber/authenticators_gcp/features/authn_gcf_token_errors.feature (hunk 8 above)
Comment (eranha) [reply cucumber/authenticators_gcp/features/authn_gcf_token_errors.feature:26]: How do I remove annotation?

File: cucumber/authenticators_gcp/features/authn_gcf.feature (hunk 7 above)
Comment (eranha) [reply cucumber/authenticators_gcp/features/authn_gcf.feature:1]: label?

Comment (eladkug) [reply cucumber/authenticators_gcp/features/authn_gcf.feature:1]: This is an example
image

Comment (nessiLahav) [reply cucumber/authenticators_gcp/features/authn_gcf.feature:1]: @eladkug your suggestion is confusing, because the authenticator is GCP, but tokens are from GCE, 
Maybe:
GCP Authenticator - GCE flow, Hosts can authenticate with GCP authenticator

Or

GCP Authenticator - Hosts can authenticate GCP authenticator with GCE token

File: cucumber/authenticators_gcp/features/authn_gcf_token_errors.feature (hunk 8 above)
Comment (nessiLahav) [reply cucumber/authenticators_gcp/features/authn_gcf_token_errors.feature:26]: see previous comments i mentioned a step there

File: dev/cli [hunk 9]
```diff
@@ -245,4 +266,4 @@ while true ; do
     key ) docker-compose exec -T conjur conjurctl role retrieve-key cucumber:user:admin ; shift ;;
      * ) if [ -z "$1" ]; then break; else echo "$1 is not a valid option"; exit 1; fi;;
   esac
-done
+done
```
Comment (nessiLahav) [inline dev/cli:269]: missing mew line

**Comment by nessiLahav** [comment]
Delete all the tokens from the PR

File: dev/cli [hunk 10]
```diff
@@ -47,7 +47,7 @@ GLOBAL OPTIONS
     --help                                    - Show this message
 COMMANDS
     -gce     Google compute engine instance name
-    -gcf     Google cloud function URL (Predefine ci/authn-gcp/function/main.py code as GCF)
```
Comment (nessiLahav) [inline dev/cli:50]: You have some mistakes in this merge, you took the old version of the file

File: dev/cli [hunk 11]
```diff
@@ -64,10 +64,6 @@ function validate_gcp_options_and_args() {
   if [ "$1" == "-h" ] || [ "$1" == "--help" ]; then
     print_gcp_help
     exit 1
-  elif [ -z "$1" ];  then
```
Comment (nessiLahav) [inline dev/cli:67]: also here, an old code version

File: dev/cli [hunk 12]
```diff
@@ -87,7 +83,7 @@ function validate_gcp_option() {
   local option_arg=$2
   case "$option" in
     -gce | -gcf ) validate_gcp_arguments_exist $option $option_arg ;;
-    * ) echo "$option is not a valid option"; print_gcp_help; exit 1;;
```
Comment (nessiLahav) [inline dev/cli:90]: also here, an old code version

File: dev/cli [hunk 13]
```diff
@@ -120,34 +123,36 @@ function enable_gcp() {
 function fetch_gce_tokens() {
   local gce_instance_name="$1"
 
- "../ci/authn-gcp/get_gcp_id_tokens.sh" "$gce_instance_name"
-
-  gce_token_file="../ci/authn-gcp/tokens/gce_token_valid"
-
-  if ! [ -f "$gce_token_file" ]; then
-    echo "GCP token file $gce_token_file does not exist."
+  if [ -z "$gce_instance_name" ]; then
```
Comment (nessiLahav) [inline dev/cli:126]: also here, an old code version, this is redundant now i already check it

File: dev/cli [hunk 14]
```diff
@@ -166,21 +171,37 @@ function fetch_gcf_tokens() {
...
 
 function validate_gcf_url_accessible() {
   echo "Validating google cloud function is accessible"
+
   local gcp_func_url=$1
   local audience="dummy_host"
+  local status_code=$(curl -o /dev/null -s -w "%{http_code}\n" \
+    -H 'Metadata-Flavor: Google' "$gcp_func_url?audience=$audience")
 
-  sh ../ci/authn-gcp/validate_gcf_url_accessible.sh "$gcp_func_url" "$audience" || exit 1
```
Comment (nessiLahav) [inline dev/cli:183]: also here, an old code version

File: cucumber/authenticators_gcp/features/authn_gcf.feature (hunk 7 above)
Comment (eladkug) [reply cucumber/authenticators_gcp/features/authn_gcf.feature:1]: i'm sorry , it was example for the label.
GCP Authenticator - GCE flow, Hosts can authenticate with GCP authenticator --> look good

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb [hunk 15]
```diff
@@ -24,19 +24,83 @@
   set_annotation_to_resource("authn-gcp/#{annotation_name}", annotation_value)
 end
 
-# Sets all GCP annotations
+Given(/^I set "authn-gcp\/(service-account-id|service-account-email|project-id|instance-name)" annotation to function host "([^"]*)"$/) do |annotation_name, hostname|
```
Comment (orenbm) [inline cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:27]: why is this step duplicated? how is it different from the step above besides the word `function`?

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb [hunk 16]
```diff
@@ -24,19 +24,83 @@
...
+          when "no"
+            nil
+          when "empty"
+            ""
+          when "self signed"
+            self_signed_token
+          when "no kid"
+            no_kid_self_signed_token
+          else
+            @gcf_identity_token
```
Comment (orenbm) [inline cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:96]: can this step be combined with GCE and return `@gcf_identity_token` or `@gce_identity_token` according to the presence of the word `function`?

Comment (orenbm) [reply cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:96]: same question for other steps - can we combine some of them to have less code?

File: dev/cli (hunk 10 above)
Comment (orenbm) [reply dev/cli:50]: why is this `-gce` and not `--gce`? same for `gcf`

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb (hunk 16 above)
Comment (eranha) [reply cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:96]: it's possible but function uses only valid token so I simplified it more.

File: dev/cli (hunk 10 above)
Comment (nessiLahav) [reply dev/cli:50]: because is a sub parameter for `--authn-gcp` so i wanted is to be different , but i dont have s storing opinion it could be also `--gce` and `--gcf`

File: dev/cli [hunk 17]
```diff
@@ -100,6 +100,13 @@ function validate_gcp_arguments_exist() {
   fi
 }
 
+function print_gcp_option_error() {
```
Comment (nessiLahav) [inline dev/cli:103]: delete this function

File: cucumber/authenticators_gcp/features/authn_gcf.feature [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    And The following appears in the audit log after my savepoint:
+    """
+    cucumber:host:test-app successfully authenticated with authenticator authn-gcp service cucumber:webservice:conjur/authn-gcp
+    """
+
+  Scenario: Hosts can authenticate with GCP authenticator using service-account-email annotation only and fetch secret
+    Given I have a "variable" resource called "test-variable"
+    And I add the secret value "test-secret" to the resource "cucumber:variable:test-variable"
+    And I permit host "test-app" to "execute" it
+    And I set service-account-email GCF annotations to host "test-app"
```
Comment (nessiLahav) [inline cucumber/authenticators_gcp/features/authn_gcf.feature:63]: Add step of removing annoations

File: cucumber/authenticators_gcp/features/authn_gcf.feature [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+    """
+    cucumber:host:test-app successfully authenticated with authenticator authn-gcp service cucumber:webservice:conjur/authn-gcp
+    """
+
+  Scenario: Hosts can authenticate with GCP authenticator using service-account-id annotation only and fetch secret
+    Given I have a "variable" resource called "test-variable"
+    And I add the secret value "test-secret" to the resource "cucumber:variable:test-variable"
+    And I permit host "test-app" to "execute" it
+    And I remove all annotations from host "test-app"
+    And I set service-account-id GCF annotations to host "test-app"
```
Comment (nessiLahav) [inline cucumber/authenticators_gcp/features/authn_gcf.feature:48]: Why GCF and GCE steps looks different ? 
GCF: `And I set service-account-id GCF annotations to host "test-app"`
GCE: `And I set "authn-gcp/service-account-email" annotation to host "test-app"`

Please change it, a possible solution can be:
GCF: `And I set "authn-gcp/service-account-email" annotation with valid GCF value to host "test-app"`
GCE: `And I set "authn-gcp/service-account-email" annotation with valid GCE value to host "test-app"`

File: cucumber/authenticators_gcp/features/authn_gcf_token_errors.feature [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+    And I obtain a valid GCF identity token
+    And I save my place in the log file
+    When I authenticate with authn-gcp using a valid GCF identity token
+    Then it is unauthorized
+    And The following appears in the log after my savepoint:
+    """
+    CONJ00049E Resource restriction 'authn-gcp/service-account-id' does not match resource in JWT token
+    """
+
+  Scenario: Token with project-id host annotation is denied
```
Comment (eladkug) [inline cucumber/authenticators_gcp/features/authn_gcf_token_errors.feature:75]: what happen if i declare service-account-email annotation and project-id annotation ? it suppose to work, but since they are going on the same path we need to verify it will not failed on project-id not exist.

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb [hunk 21]
```diff
@@ -24,19 +24,68 @@
...
+end
+
 Given(/^I set all valid GCP annotations to (user|host) "([^"]*)"$/) do |role_type, hostname|
   i_have_a_resource role_type, hostname
 
-  set_annotation_to_resource("authn-gcp/service-account-id", gcp_service_account_id)
-  set_annotation_to_resource("authn-gcp/service-account-email", gcp_service_account_email)
-  set_annotation_to_resource("authn-gcp/project-id", gcp_project_id)
-  set_annotation_to_resource("authn-gcp/instance-name", gcp_instance_name)
+  set_annotation_to_resource("authn-gcp/service-account-id", gce_service_account_id)
```
Comment (eladkug) [inline cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:54]: its a little bit confusing:
Given **GCP** annotations
**gce_service_account** which relevant for gce/gcf
**project-id** which relevant for gce only
can we declare something like : gce_gcf_servce_account_id , gce_project_id or change to gcp_service_account_id so we will know its for both.

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb [hunk 22]
```diff
@@ -24,19 +24,68 @@
...
+
+  set_annotation_to_resource("authn-gcp/#{annotation_name}", annotation_value)
+end
+
+Given(/^I set "authn-gcp\/(service-account-id|service-account-email)" annotation with value: "([^"]*)" to host "([^"]*)"$/) do |annotation_name, annotation_value, hostname|
+  i_have_a_resource "host", hostname
+  set_annotation_to_resource("authn-gcp/#{annotation_name}", annotation_value)
+end
+
 Given(/^I set all valid GCP annotations to (user|host) "([^"]*)"$/) do |role_type, hostname|
```
Comment (eladkug) [inline cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:51]: GCP need to be GCE

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb [hunk 23]
```diff
@@ -24,19 +24,68 @@
...
+    token_type.to_sym
+  )
+end
+
+Given(/I authenticate with authn-gcp using a valid GCF identity token/) do
+  authenticate_gcp_token(account: AuthnGcpHelper::ACCOUNT, gcp_token: @gcf_identity_token)
 end
 
-# Runs a curl command in a remote machine
 Given(/^I obtain an? (valid|standard_format|user_audience|invalid_audience|non_existing_host|non_rooted_host|non_existing_account) GCP identity token$/) do |token_type|
```
Comment (eladkug) [inline cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:87]: GCE

File: cucumber/authenticators_gcp/features/authn_gcf_token_errors.feature (hunk 20 above)
Comment (nessiLahav) [reply cucumber/authenticators_gcp/features/authn_gcf_token_errors.feature:75]: @eladkug i think test 13 under [error handling ](https://ca-il-confluence.il.cyber-ark.com/display/rndp/Conjur+GCP+authenticator+2nd+phase+-+Test+plan#ConjurGCPauthenticator2ndphase-Testplan-Errorhandling)cover this use case, WDYT?

Comment (eladkug) [reply cucumber/authenticators_gcp/features/authn_gcf_token_errors.feature:75]: Yes, but i want it on automation as well.

**Review by nessiLahav — CHANGES_REQUESTED** [review]
There are 2 missing tests from the test plan with empty annotation values

**Review by nessiLahav — APPROVED** [review]
LGTM

File: dev/cli [hunk 24]
```diff
@@ -166,11 +163,21 @@ function fetch_gcf_tokens() {
...
   cd ../ci/authn-gcp/
   sh ./get_tokens_to_files.sh "$gcp_func_url" "$token_prefix" || exit 1
   cd -
+  local decoded_gcf_token_payload=$(_get_gcp_token_payload "../ci/authn-gcp/tokens/gcf_token_valid")
+
+  echo "Setting GCF details as env variables"
+  local gcf_service_account_id="$(echo "$decoded_gcf_token_payload" | jq -r '.sub')"
+  local gcf_service_account_email="$(echo "$decoded_gcf_token_payload" | jq -r '.email')"
+
+  local gce_env_args="-e GCF_SERVICE_ACCOUNT_ID=$gcf_service_account_id \
```
Comment (nessiLahav) [inline dev/cli:177]: should be `gcf_env_args`
